File: packages/blitsen/test/native-harness/web-socket.mjs
// `WebSocket` (issue #80), against a real server on a real socket.
//
// The server is hand-rolled on `node:net` rather than taken from `Bun.serve`
// for the reason the fetch probe uses `node:http`: `runBridgeHarness` installs
// the bridge into *this* context, so `Response`, `Blob` and the rest are
// Blitsen's by the time a server handler would run. A raw socket has no such
// surface to be handed. RFC 6455 is small enough at this scale — a handshake,
// and frames a browser client always masks and a server never does.
import { strict as assert } from "node:assert";
import { createHash } from "node:crypto";
import { createServer } from "node:net";

import { native } from "./addon.mjs";

const GUID = "258EAFA5-E914-47DA-95CA-C5AB0DC85B11";
const accept = key => createHash("sha1").update(key + GUID).digest("base64");

/** Frames a payload the way a server does: never masked. */
const frame = (opcode, payload) => {
  const body = Buffer.isBuffer(payload) ? payload : Buffer.from(payload, "utf8");
  const header = body.length < 126 ? Buffer.from([0x80 | opcode, body.length])
    : body.length < 65536
      ? Buffer.concat([Buffer.from([0x80 | opcode, 126]), (() => {
        const length = Buffer.alloc(2); length.writeUInt16BE(body.length); return length;
      })()])
      : Buffer.concat([Buffer.from([0x80 | opcode, 127]), (() => {
        const length = Buffer.alloc(8); length.writeBigUInt64BE(BigInt(body.length)); return length;
      })()]);
  return Buffer.concat([header, body]);
};

/**
 * Pulls one complete frame off the front of `buffer`.
 *
 * Returns null when the frame is not all there yet, which on a stream socket is
 * an ordinary state rather than an error.
 */
const readFrame = buffer => {
  if (buffer.length < 2) return null;
  const opcode = buffer[0] & 0x0f;
  const masked = (buffer[1] & 0x80) !== 0;
  let length = buffer[1] & 0x7f;
  let offset = 2;
  if (length === 126) {
    if (buffer.length < 4) return null;
    length = buffer.readUInt16BE(2);
    offset = 4;
  } else if (length === 127) {
    if (buffer.length < 10) return null;
    length = Number(buffer.readBigUInt64BE(2));
    offset = 10;
  }
  const mask = masked ? buffer.subarray(offset, offset + 4) : null;
  if (masked) offset += 4;
  if (buffer.length < offset + length) return null;
  const payload = Buffer.from(buffer.subarray(offset, offset + length));
  // A client must mask every frame it sends, so unmasking is not optional.
  if (mask) for (let index = 0; index < payload.length; index += 1) payload[index] ^= mask[index % 4];
  return { opcode, payload, rest: buffer.subarray(offset + length) };
};

// What the server does with what it is sent, so the client's half can be
// asserted from the outside as well as from inside the runtime.
const received = [];
let negotiated = null;

const server = createServer(connection => {
  let buffer = Buffer.alloc(0);
  let open = false;
  connection.on("data", chunk => {
    buffer = Buffer.concat([buffer, chunk]);
    if (!open) {
      const end = buffer.indexOf("\r\n\r\n");
      if (end < 0) return;
      const request = buffer.subarray(0, end).toString();
      buffer = buffer.subarray(end + 4);
      const key = /sec-websocket-key:\s*(\S+)/i.exec(request)?.[1];
      // The client asks for two and the server picks the second, so the test
      // can tell a negotiated protocol from an echoed one.
      const offered = /sec-websocket-protocol:\s*(.+)/i.exec(request)?.[1]
        ?.split(",").map(value => value.trim()) ?? [];
      negotiated = offered[1] ?? offered[0] ?? null;
      connection.write([
        "HTTP/1.1 101 Switching Protocols",
        "Upgrade: websocket",
        "Connection: Upgrade",
        `Sec-WebSocket-Accept: ${accept(key)}`,
        ...(negotiated ? [`Sec-WebSocket-Protocol: ${negotiated}`] : []),
        "", "",
      ].join("\r\n"));
      open = true;
      // Both frame kinds, unprompted, so the client's `binaryType` handling is
      // exercised on data it did not ask for at a moment it chose.
      connection.write(frame(0x1, "hello"));
      connection.write(frame(0x2, Buffer.from([1, 2, 3, 4])));
    }
    for (let parsed = readFrame(buffer); parsed; parsed = readFrame(buffer)) {
      buffer = parsed.rest;
      if (parsed.opcode === 0x8) {
        const code = parsed.payload.length >= 2 ? parsed.payload.readUInt16BE(0) : null;
        received.push(["close", code, parsed.payload.subarray(2).toString()]);
        connection.end(frame(0x8, parsed.payload));
      } else if (parsed.opcode === 0x1) {
        received.push(["text", parsed.payload.toString()]);
        // The echo is what proves a text frame made the round trip.
        connection.write(frame(0x1, `echo:${parsed.payload.toString()}`));
      } else if (parsed.opcode === 0x2) {
        received.push(["binary", [...parsed.payload]]);
      }
    }
  });
  connection.on("error", () => {});
});
await new Promise(resolve => server.listen(0, "127.0.0.1", resolve));
const address = `ws://127.0.0.1:${server.address().port}/socket`;

try {
  const opened = JSON.parse(native.runBridgeHarness(
    `<div id="socket"></div>`,
    `{ const results = globalThis.__blitsenSocket = { events: [], errors: [] };
       if (!("WebSocket" in globalThis)) throw new Error("WebSocket must be installed");
       // The readyState constants live on the constructor and on instances.
       if (WebSocket.CONNECTING !== 0 || WebSocket.OPEN !== 1 || WebSocket.CLOSING !== 2
           || WebSocket.CLOSED !== 3) throw new Error("readyState constants");

       for (const [address, expected] of [["http://example.com/", "SyntaxError"],
         ["not a url", "SyntaxError"]]) {
         let refused = null;
         try { new WebSocket(address); } catch (error) { refused = error.name; }
         if (refused !== expected) throw new Error("a non-ws address must be refused: " + address);
       }

       const socket = globalThis.__blitsenSocketHandle =
         new WebSocket("${address}", ["chat.v1", "chat.v2"]);
       if (socket.readyState !== WebSocket.CONNECTING) throw new Error("a new socket is CONNECTING");
       if (socket.url !== "${address}") throw new Error("url reads back what it was given");
       if (socket.bufferedAmount !== 0) throw new Error("nothing is buffered yet");
       if (socket.binaryType !== "blob") throw new Error("binaryType defaults to blob");
       let refusedType = null;
       try { socket.binaryType = "text"; } catch (error) { refusedType = error.constructor.name; }
       if (refusedType !== "TypeError") throw new Error("an unknown binaryType is refused");
       socket.binaryType = "arraybuffer";

       // Sending before the socket is open is the one thing that throws.
       let early = null;
       try { socket.send("too soon"); } catch (error) { early = error.name; }
       if (early !== "InvalidStateError") throw new Error("send before open must throw");

       socket.addEventListener("open", () => {
         results.events.push(["open", socket.readyState, socket.protocol]);
         socket.send("ping");
         socket.send(new Uint8Array([9, 8, 7]).buffer);
       });
       socket.addEventListener("message", event => {
         const data = event.data;
         if (typeof data === "string") results.events.push(["text", data]);
         else if (data instanceof ArrayBuffer) results.events.push(["binary", [...new Uint8Array(data)]]);
         else results.events.push(["other", String(data)]);
         if (results.events.filter(entry => entry[0] === "text").length === 2)
           socket.close(4000, "done here");
       });
       socket.addEventListener("close", event => {
         results.events.push(["close", event.code, event.reason, event.wasClean, socket.readyState]);
         results.done = true;
       });
       socket.addEventListener("error", () => results.errors.push("error"));
       document.getElementById("socket").setAttribute("data-socket", "ok"); }`,
    320, 180));
  assert.equal(
    opened.nodes.find(node => node.attributes.id === "socket").attributes["data-socket"], "ok");

  // The same landing rule the network results follow: a frame that arrived off
  // the socket waits for the frame turn rather than interrupting between turns.
  await Bun.sleep(250);
  assert.equal(globalThis.__blitsenSocket.events.length, 0,
    "socket events wait for the frame turn rather than arriving between them");
  assert.equal(globalThis.__blitsenAnimationFramesPending(), true,
    "a live socket keeps the host turning so its frames can land");

  for (let turn = 0; turn < 400 && !globalThis.__blitsenSocket.done; turn++) {
    globalThis.__blitsenAnimationFrameTick(0);
    await Bun.sleep(5);
  }
  const { events, errors, done } = globalThis.__blitsenSocket;
  assert.equal(done, true, "the socket closed within the turns allowed");
  assert.deepEqual(errors, [], "a clean session reports no error");

  const byKind = kind => events.filter(entry => entry[0] === kind);
  assert.deepEqual(byKind("open")[0], ["open", 1, "chat.v2"],
    "open reports OPEN and the protocol the server chose, not the first offered");
  assert.deepEqual(byKind("text").map(entry => entry[1]), ["hello", "echo:ping"],
    "text frames arrive as strings, in order");
  assert.deepEqual(byKind("binary")[0], ["binary", [1, 2, 3, 4]],
    "a binary frame arrives as an ArrayBuffer under binaryType arraybuffer");
  assert.deepEqual(byKind("close")[0], ["close", 4000, "done here", true, 3],
    "close carries its code and reason, is clean, and leaves the socket CLOSED");

  // The other half: what the server actually received off the wire.
  assert.deepEqual(received.find(entry => entry[0] === "text"), ["text", "ping"]);
  assert.deepEqual(received.find(entry => entry[0] === "binary"), ["binary", [9, 8, 7]]);
  assert.deepEqual(received.find(entry => entry[0] === "close"), ["close", 4000, "done here"]);
  assert.equal(negotiated, "chat.v2");
  assert.equal(globalThis.__blitsenAnimationFramesPending(), false,
    "a closed socket stops asking for frames");

  // Blob is the default, and is what an application that never sets
  // `binaryType` gets. Checked in its own session so the first one can assert
  // the arraybuffer path without branching.
  received.length = 0;
  JSON.parse(native.runBridgeHarness(
    `<div id="blob"></div>`,
    `{ const results = globalThis.__blitsenBlobSocket = { seen: [] };
       const socket = new WebSocket("${address}");
       socket.addEventListener("message", event => {
         if (event.data instanceof Blob) {
           results.seen.push(["blob", event.data.size, event.data.type]);
           socket.close();
         }
       });
       socket.addEventListener("close", () => { results.done = true; }); }`,
    320, 180));
  for (let turn = 0; turn < 400 && !globalThis.__blitsenBlobSocket.done; turn++) {
    globalThis.__blitsenAnimationFrameTick(0);
    await Bun.sleep(5);
  }
  assert.deepEqual(globalThis.__blitsenBlobSocket.seen[0], ["blob", 4, ""],
    "the default binaryType delivers a Blob of the frame's bytes");

  delete globalThis.__blitsenSocket;
  delete globalThis.__blitsenSocketHandle;
  delete globalThis.__blitsenBlobSocket;
} finally {
  server.close();
}

File: crates/blitsen-host/src/dom_bridge/web_socket.rs
//! Off-main-thread WebSocket connections for the DOM bridge.
//!
//! A connection is one task on the shared worker pool, never on the thread that
//! owns the DOM. Everything it observes — the handshake, each frame, the close —
//! queues in [`Shared`] and JavaScript drains it at exactly one point in the
//! frame turn, the start of the animation-frame stage, before any
//! `requestAnimationFrame` callback runs. A message therefore cannot arrive
//! part-way through a frame, which is the same contract `fetch` completions and
//! dialog answers are delivered under.
//!
//! Binary payloads stay in Rust until the delivering turn asks for them, keyed by
//! socket and arrival order, so the bytes cross the engine boundary once and in
//! the shape `binaryType` asked for rather than through a string.

use std::collections::HashMap;
use std::sync::atomic::{AtomicU64, AtomicUsize, Ordering};
use std::sync::{Arc, Mutex};

use blitsen_js::JsError;
use futures_util::{SinkExt, StreamExt};
use serde_json::{Value, json};
use tokio::runtime::Runtime;
use tokio::sync::mpsc::{self, UnboundedReceiver, UnboundedSender};
use tokio_tungstenite::connect_async;
use tokio_tungstenite::tungstenite::client::IntoClientRequest;
use tokio_tungstenite::tungstenite::http::HeaderValue;
use tokio_tungstenite::tungstenite::http::header::SEC_WEBSOCKET_PROTOCOL;
use tokio_tungstenite::tungstenite::protocol::frame::coding::CloseCode;
use tokio_tungstenite::tungstenite::{Bytes, Message, Utf8Bytes, protocol::CloseFrame};
use url::Url;

use super::net_pool::{lock, runtime as net_runtime};

/// Reported when a connection ends without a close frame from either side.
const ABNORMAL_CLOSURE: u16 = 1006;
/// Reported when the close handshake completed but carried no status.
const NO_STATUS_RECEIVED: u16 = 1005;

/// What the frame turn hands back to a running connection.
enum Command {
    /// The byte count is carried so the buffered total can be released once the
    /// transport has taken the message, rather than when it was queued.
    Send {
        message: Message,
        bytes: usize,
    },
    Close(Option<CloseFrame>),
}

/// State shared with the worker pool.
#[derive(Default)]
struct Shared {
    events: Mutex<Vec<Value>>,
    payloads: Mutex<HashMap<(u64, u64), Vec<u8>>>,
}

impl Shared {
    fn push(&self, event: Value) {
        lock(&self.events).push(event);
    }

    /// Ends a connection the way the spec's "fail the WebSocket connection"
    /// does: an `error` the application can see, then a close it did not choose.
    fn fail(&self, id: u64) {
        self.push(json!({ "id": id, "type": "error" }));
        self.push(json!({
            "id": id, "type": "close",
            "code": ABNORMAL_CLOSURE, "reason": "", "wasClean": false,
        }));
    }
}

/// One live connection, from the frame turn's side.
struct Connection {
    commands: UnboundedSender<Command>,
    buffered: Arc<AtomicUsize>,
    task: tokio::task::AbortHandle,
}

/// The WebSocket executor owned by one JavaScript context.
pub(super) struct WebSocketHost {
    runtime: &'static Runtime,
    next_id: AtomicU64,
    open: Mutex<HashMap<u64, Connection>>,
    shared: Arc<Shared>,
}

/// Reads the subprotocol the server agreed to, which is "" when it agreed none.
fn negotiated_protocol(headers: &tokio_tungstenite::tungstenite::http::HeaderMap) -> String {
    headers
        .get(SEC_WEBSOCKET_PROTOCOL)
        .and_then(|value| value.to_str().ok())
        .unwrap_or_default()
        .to_owned()
}

/// Runs one connection to its close, queueing everything it observes.
async fn run(
    id: u64,
    request: tokio_tungstenite::tungstenite::handshake::client::Request,
    shared: Arc<Shared>,
    buffered: Arc<AtomicUsize>,
    mut commands: UnboundedReceiver<Command>,
) {
    let connected = tokio::select! {
        result = connect_async(request) => result,
        // Only a close can arrive before the socket is open — `send` is refused
        // while the connection is opening — and the spec answers it by failing
        // the connection rather than by completing a handshake it will discard.
        _ = commands.recv() => {
            shared.fail(id);
            return;
        }
    };
    let (socket, response) = match connected {
        Ok(connected) => connected,
        Err(error) => {
            // The message is queued rather than thrown: a handshake that fails
            // is a `close` the application listens for, not a constructor error.
            shared.push(json!({ "id": id, "type": "error", "message": error.to_string() }));
            shared.push(json!({
                "id": id, "type": "close",
                "code": ABNORMAL_CLOSURE, "reason": "", "wasClean": false,
            }));
            return;
        }
    };
    shared.push(json!({
        "id": id, "type": "open", "protocol": negotiated_protocol(response.headers()),
    }));

    let (mut sink, mut stream) = socket.split();
    let mut sequence = 0_u64;
    let mut closing = false;
    let mut received: Option<CloseFrame> = None;
    let clean = loop {
        tokio::select! {
            // Biased so a queued frame is written before the next read is
            // awaited: a send issued in frame N must not sit behind a peer that
            // has gone quiet.
            biased;
            command = commands.recv(), if !closing => match command {
                Some(Command::Send { message, bytes }) => {
                    let sent = sink.send(message).await;
                    buffered.fetch_sub(bytes, Ordering::Relaxed);
                    if sent.is_err() { break false; }
                }
                // `None` is the host being disposed of, which drops the sender.
                command => {
                    let frame = match command {
                        Some(Command::Close(frame)) => frame,
                        _ => None,
                    };
                    closing = true;
                    if sink.send(Message::Close(frame)).await.is_err() { break false; }
                }
            },
            message = stream.next() => match message {
                // The stream ends once the close handshake has been echoed both
                // ways, which is the one exit that is a clean close.
                None => break true,
                Some(Err(_)) => break false,
                Some(Ok(Message::Text(text))) => shared.push(json!({
                    "id": id, "type": "message", "text": text.as_str(),
                })),
                Some(Ok(Message::Binary(bytes))) => {
                    sequence += 1;
                    lock(&shared.payloads).insert((id, sequence), bytes.to_vec());
                    shared.push(json!({ "id": id, "type": "message", "binary": sequence }));
                }
                // The peer closed first. Reading on lets the protocol send the
                // reply, after which the stream ends and the close is clean.
                Some(Ok(Message::Close(frame))) => {
                    received = frame;
                    closing = true;
                }
                // Ping, pong and raw frames are the protocol's own business:
                // tungstenite answers a ping itself and nothing is observable.
                Some(Ok(_)) => {}
            },
        }
    };
    if !clean {
        shared.push(json!({ "id": id, "type": "error" }));
    }
    let (code, reason) = match received {
        Some(frame) => (u16::from(frame.code), frame.reason.as_str().to_owned()),
        None if clean => (NO_STATUS_RECEIVED, String::new()),
        None => (ABNORMAL_CLOSURE, String::new()),
    };
    shared.push(json!({
        "id": id, "type": "close", "code": code, "reason": reason, "wasClean": clean,
    }));
}

impl WebSocketHost {
    /// Creates a host bound to the shared worker pool.
    pub(super) fn new() -> Result<Self, JsError> {
        Ok(Self {
            runtime: net_runtime()?,
            next_id: AtomicU64::new(1),
            open: Mutex::new(HashMap::new()),
            shared: Arc::default(),
        })
    }

    /// Opens a connection on the worker pool and returns its identifier.
    ///
    /// Only the address is refused here. Everything else a handshake can go
    /// wrong about is reported as `error` then `close`, because that is what the
    /// application has a listener for.
    pub(super) fn open(&self, url: &str, protocols: &[String]) -> Result<u64, JsError> {
        let parsed = Url::parse(url)
            .map_err(|error| JsError::new(format!("invalid WebSocket URL {url}: {error}")))?;
        if !matches!(parsed.scheme(), "ws" | "wss") {
            return Err(JsError::new(format!(
                "a WebSocket address is ws: or wss:; {}: is not",
                parsed.scheme()
            )));
        }
        let mut request = parsed
            .as_str()
            .into_client_request()
            .map_err(|error| JsError::new(format!("invalid WebSocket URL {url}: {error}")))?;
        if !protocols.is_empty() {
            let value = HeaderValue::from_str(&protocols.join(", "))
                .map_err(|_| JsError::new("a WebSocket subprotocol must be a header-safe token"))?;
            request.headers_mut().insert(SEC_WEBSOCKET_PROTOCOL, value);
        }
        let id = self.next_id.fetch_add(1, Ordering::Relaxed);
        let (commands, receiver) = mpsc::unbounded_channel();
        let buffered = Arc::new(AtomicUsize::new(0));
        let shared = Arc::clone(&self.shared);
        let task = self
            .runtime
            .spawn(run(id, request, shared, Arc::clone(&buffered), receiver));
        lock(&self.open).insert(
            id,
            Connection {
                commands,
                buffered,
                task: task.abort_handle(),
            },
        );
        Ok(id)
    }

    /// Queues a message, counting its bytes against the buffered total.
    ///
    /// A socket that has already gone is not an error: JavaScript refuses a send
    /// on a closed socket itself, and the close it has not drained yet is one the
    /// spec discards the message for.
    fn queue(&self, id: u64, message: Message, bytes: usize) {
        let open = lock(&self.open);
        let Some(connection) = open.get(&id) else {
            return;
        };
        connection.buffered.fetch_add(bytes, Ordering::Relaxed);
        if connection
            .commands
            .send(Command::Send { message, bytes })
            .is_err()
        {
            connection.buffered.fetch_sub(bytes, Ordering::Relaxed);
        }
    }

    pub(super) fn send_text(&self, id: u64, text: String) {
        let bytes = text.len();
        self.queue(id, Message::Text(Utf8Bytes::from(text)), bytes);
    }

    pub(super) fn send_binary(&self, id: u64, payload: Vec<u8>) {
        let bytes = payload.len();
        self.queue(id, Message::Binary(Bytes::from(payload)), bytes);
    }

    /// Bytes queued and not yet handed to the transport.
    pub(super) fn buffered(&self, id: u64) -> usize {
        lock(&self.open)
            .get(&id)
            .map_or(0, |connection| connection.buffered.load(Ordering::Relaxed))
    }

    /// Starts the close handshake. The `close` event follows from the worker.
    pub(super) fn close(&self, id: u64, code: Option<u16>, reason: &str) {
        let open = lock(&self.open);
        let Some(connection) = open.get(&id) else {
            return;
        };
        let frame = code.map(|code| CloseFrame {
            code: CloseCode::from(code),
            reason: Utf8Bytes::from(reason.to_owned()),
        });
        let _ = connection.commands.send(Command::Close(frame));
    }

    /// Drains everything the connections observed since the previous frame turn.
    pub(super) fn poll(&self) -> Value {
        let events = std::mem::take(&mut *lock(&self.shared.events));
        let mut open = lock(&self.open);
        for event in &events {
            if event["type"] == "close"
                && let Some(id) = event["id"].as_u64()
            {
                open.remove(&id);
            }
        }
        Value::Array(events)
    }

    /// Takes a binary message's bytes, which are handed over exactly once.
    pub(super) fn take_binary(&self, id: u64, sequence: u64) -> Result<Vec<u8>, JsError> {
        lock(&self.shared.payloads)
            .remove(&(id, sequence))
            .ok_or_else(|| JsError::new("the WebSocket message is no longer available"))
    }

    /// Drops every connection and everything they had queued.
    pub(super) fn dispose(&self) {
        for (_, connection) in lock(&self.open).drain() {
            connection.task.abort();
        }
        lock(&self.shared.events).clear();
        lock(&self.shared.payloads).clear();
    }
}

#[cfg(test)]
mod tests {
    use std::net::TcpListener;

    use super::*;

    /// Accepts one connection and runs the exchange the test asked for.
    ///
    /// The server is tungstenite's own accept side on a thread of its own, so
    /// what the host is measured against is a real handshake and real framing.
    ///
    /// The handshake callback's error type is tungstenite's `ErrorResponse`, an
    /// `http::Response` the `Callback` signature fixes, so the large-`Err` lint
    /// has nothing here to act on.
    #[allow(clippy::result_large_err)]
    fn server(
        exchange: impl FnOnce(&mut tokio_tungstenite::tungstenite::WebSocket<std::net::TcpStream>)
        + Send
        + 'static,
    ) -> (String, std::thread::JoinHandle<()>) {
        let listener = TcpListener::bind("127.0.0.1:0").unwrap();
        let url = format!("ws://{}/socket", listener.local_addr().unwrap());
        let handle = std::thread::spawn(move || {
            let (stream, _) = listener.accept().unwrap();
            let mut socket = tokio_tungstenite::tungstenite::accept_hdr(
                stream,
                |request: &tokio_tungstenite::tungstenite::handshake::server::Request,
                 mut response: tokio_tungstenite::tungstenite::handshake::server::Response| {
                    // Echo the first subprotocol offered, which is what a server
                    // that supports one of them does.
                    if let Some(offered) = request.headers().get(SEC_WEBSOCKET_PROTOCOL)
                        && let Ok(offered) = offered.to_str()
                        && let Some(first) = offered.split(',').next()
                    {
                        response.headers_mut().insert(
                            SEC_WEBSOCKET_PROTOCOL,
                            HeaderValue::from_str(first.trim()).unwrap(),
                        );
                    }
                    Ok(response)
                },
            )
            .unwrap();
            exchange(&mut socket);
        });
        (url, handle)
    }

    /// Turns the frame loop's delivery point by hand until the events land.
    fn drain(host: &WebSocketHost, until: &str) -> Vec<Value> {
        let mut events = Vec::new();
        for _ in 0..600 {
            if let Value::Array(polled) = host.poll() {
                events.extend(polled);
            }
            if events.iter().any(|event| event["type"] == until) {
                return events;
            }
            std::thread::sleep(std::time::Duration::from_millis(10));
        }
        panic!("no {until} event arrived: {events:?}");
    }

    #[test]
    fn a_handshake_frames_and_a_clean_close_arrive_on_the_queue_in_order() {
        let (url, server) = server(|socket| {
            let text = socket.read().unwrap();
            socket.send(text).unwrap();
            socket
                .send(Message::Binary(Bytes::from_static(&[7, 8, 9])))
                .unwrap();
            // Reads until the client's close, which tungstenite answers itself.
            while socket.read().is_ok() {}
        });
        let host = WebSocketHost::new().unwrap();
        let id = host.open(&url, &["chat.v1".to_owned()]).unwrap();
        let opened = drain(&host, "open");
        assert_eq!(opened[0]["type"], "open");
        assert_eq!(opened[0]["protocol"], "chat.v1");

        host.send_text(id, "hello".to_owned());
        let mut events = drain(&host, "message");
        while events
            .iter()
            .filter(|event| event["type"] == "message")
            .count()
            < 2
        {
            events.extend(drain(&host, "message"));
        }
        let messages = events
            .iter()
            .filter(|event| event["type"] == "message")
            .collect::<Vec<_>>();
        assert_eq!(messages[0]["text"], "hello");
        let sequence = messages[1]["binary"].as_u64().unwrap();
        assert_eq!(host.take_binary(id, sequence).unwrap(), vec![7, 8, 9]);
        assert!(
            host.take_binary(id, sequence).is_err(),
            "a payload is handed over once"
        );

        host.close(id, Some(1000), "done");
        let closed = drain(&host, "close");
        let close = closed
            .iter()
            .find(|event| event["type"] == "close")
            .unwrap();
        assert_eq!(close["code"], 1000);
        assert_eq!(close["reason"], "done");
        assert_eq!(close["wasClean"], Value::Bool(true));
        assert!(
            !closed.iter().any(|event| event["type"] == "error"),
            "a close the application asked for is not an error: {closed:?}"
        );
        assert_eq!(host.buffered(id), 0, "a closed socket buffers nothing");
        server.join().unwrap();
    }

    #[test]
    fn a_peer_that_closes_first_reports_its_own_code_and_reason() {
        let (url, server) = server(|socket| {
            socket
                .close(Some(CloseFrame {
                    code: CloseCode::from(4001),
                    reason: Utf8Bytes::from_static("go away"),
                }))
                .unwrap();
            while socket.read().is_ok() {}
        });
        let host = WebSocketHost::new().unwrap();
        host.open(&url, &[]).unwrap();
        let events = drain(&host, "close");
        let close = events
            .iter()
            .find(|event| event["type"] == "close")
            .unwrap();
        assert_eq!(close["code"], 4001);
        assert_eq!(close["reason"], "go away");
        assert_eq!(close["wasClean"], Value::Bool(true));
        server.join().unwrap();
    }

    #[test]
    fn a_handshake_that_never_completes_fails_rather_than_throwing_at_the_call() {
        let host = WebSocketHost::new().unwrap();
        // Port 1 is unroutable, so the connection fails without a server.
        let id = host.open("ws://127.0.0.1:1/socket", &[]).unwrap();
        let events = drain(&host, "close");
        assert_eq!(events[0]["type"], "error");
        assert!(events[0]["message"].is_string());
        assert_eq!(events[1]["code"], ABNORMAL_CLOSURE);
        assert_eq!(events[1]["wasClean"], Value::Bool(false));
        assert!(
            host.poll().as_array().unwrap().is_empty(),
            "a drained close forgets the connection"
        );
        assert_eq!(host.buffered(id), 0);
    }

    #[test]
    fn addresses_without_a_socket_behind_them_are_refused_at_the_call() {
        let host = WebSocketHost::new().unwrap();
        for (url, expected) in [
            ("https://example.com/socket", "is not"),
            ("blitsen://app/socket", "is not"),
            ("/relative", "invalid WebSocket URL"),
        ] {
            let error = host.open(url, &[]).unwrap_err();
            assert!(error.message().contains(expected), "{}", error.message());
        }
        let error = host
            .open("ws://127.0.0.1:1/", &["bad\nprotocol".to_owned()])
            .unwrap_err();
        assert!(
            error.message().contains("header-safe"),
            "{}",
            error.message()
        );
    }

    #[test]
    fn disposing_a_context_abandons_every_connection_it_opened() {
        let (url, server) = server(|socket| while socket.read().is_ok() {});
        let host = WebSocketHost::new().unwrap();
        let id = host.open(&url, &[]).unwrap();
        drain(&host, "open");
        host.dispose();
        assert!(host.poll().as_array().unwrap().is_empty());
        assert_eq!(host.buffered(id), 0);
        // The send lands nowhere rather than panicking on a connection that is
        // gone, which is the shape `__blitsenDisposeContext` leaves behind.
        host.send_text(id, "ignored".to_owned());
        host.close(id, Some(1000), "");
        server.join().unwrap();
    }
}
